File: standalone/ios/VPinball/utils/VPinballManager.swift
import SwiftUI

class VPinballManager {
    static let shared = VPinballManager()

    let impactGenerators: [UIImpactFeedbackGenerator.FeedbackStyle: UIImpactFeedbackGenerator] = [
        .heavy: UIImpactFeedbackGenerator(style: .heavy),
        .light: UIImpactFeedbackGenerator(style: .light),
        .medium: UIImpactFeedbackGenerator(style: .medium),
        .rigid: UIImpactFeedbackGenerator(style: .rigid),
        .soft: UIImpactFeedbackGenerator(style: .soft),
    ]

    private init() {
        impactGenerators.forEach { $0.value.prepare() }
    }

    func startup() {
        VPinballInit { value, data in
            let event = VPinballEvent(rawValue: value)
            switch event {
            case .extractScript,
                 .loadingItems,
                 .loadingSounds,
                 .loadingImages,
                 .loadingFonts,
                 .loadingCollections,
                 .prerendering:
                if let data = data {
                    let json = String(cString: UnsafePointer<CChar>(data))
                    if let jsonData = json.data(using: .utf8),
                       let progressData = try? JSONDecoder().decode(ProgressEventData.self,
                                                                    from: jsonData)
                    {
                        let progress = progressData.progress
                        let eventName = event?.name
                        Task { @MainActor in
                            if let name = eventName {
                                VPinballModel.shared.updateHUD(progress: progress,
                                                               status: name)
                            } else {
                                VPinballModel.shared.updateHUD(progress: progress)
                            }
                            CATransaction.flush()
                        }
                    }
                }
            case .playerStarted:
                Task { @MainActor in
                    VPinballModel.shared.isPlaying = true
                }
            case .rumble:
                if let data = data {
                    let json = String(cString: UnsafePointer<CChar>(data))
                    if let jsonData = json.data(using: .utf8),
                       let rumbleData = try? JSONDecoder().decode(RumbleData.self,
                                                                  from: jsonData)
                    {
                        VPinballManager.shared.rumble(rumbleData)
                    }
                }
            case .playerClosed:
                Task { @MainActor in
                    let table = VPinballModel.shared.activeTable
                    VPinballModel.shared.activeTable = nil
                    VPinballModel.shared.isPlaying = false
                    MainViewModel.shared.setAction(.stopped)

                    if let table {
                        Task {
                            TableManager.shared.clearLoadedTable(table: table)
                            await TableManager.shared.loadTables()
                            try? await Task.sleep(nanoseconds: 1_000_000_000)
                            await TableManager.shared.loadTables()
                        }
                    }
                }
            case .webServer:
                if let data = data {
                    let json = String(cString: UnsafePointer<CChar>(data))
                    if let jsonData = json.data(using: .utf8),
                       let webServerData = try? JSONDecoder().decode(WebServerData.self,
                                                                     from: jsonData)
                    {
                        Task { @MainActor in
                            VPinballModel.shared.webServerURL = webServerData.url
                        }
                    }
                } else {
                    Task { @MainActor in
                        VPinballModel.shared.webServerURL = nil
                    }
                }
            case .command:
                if let data = data {
                    let json = String(cString: UnsafePointer<CChar>(data))
                    if let jsonData = json.data(using: .utf8),
                       let commandData = try? JSONDecoder().decode(CommandData.self,
                                                                   from: jsonData)
                    {
                        if commandData.command == "reloadTables" {
                            Task {
                                await TableManager.shared.loadTables()
                            }
                        }
                    }
                }
            default:
                break
            }
        }
    }

    static func log(_ level: VPinballLogLevel, _ message: String) {
        VPinballLog(level.rawValue, message.cstring)
    }

    func loadValue(_ section: VPinballSettingsSection, _ key: String, _ defaultValue: CInt) -> CInt {
        return VPinballLoadValueInt(section.rawValue.cstring, key.cstring, defaultValue)
    }

    func loadValue(_ section: VPinballSettingsSection, _ key: String, _ defaultValue: Int) -> Int {
        return Int(loadValue(section, key, CInt(defaultValue)))
    }

    func loadValue(_ section: VPinballSettingsSection, _ key: String, _ defaultValue: Float) -> Float {
        return VPinballLoadValueFloat(section.rawValue.cstring, key.cstring, defaultValue)
    }

    func loadValue(_ section: VPinballSettingsSection, _ key: String, _ defaultValue: Bool) -> Bool {
        return VPinballLoadValueBool(section.rawValue.cstring, key.cstring, defaultValue ? 1 : 0) != 0
    }

    func loadValue(_ section: VPinballSettingsSection, _ key: String, _ defaultValue: String) -> String {
        return String(cString: VPinballLoadValueString(section.rawValue.cstring, key.cstring, defaultValue.cstring))
    }

    func saveValue(_ section: VPinballSettingsSection, _ key: String, _ value: CInt) {
        VPinballSaveValueInt(section.rawValue.cstring, key.cstring, value)
    }

    func saveValue(_ section: VPinballSettingsSection, _ key: String, _ value: Int) {
        saveValue(section, key, CInt(value))
    }

    func saveValue(_ section: VPinballSettingsSection, _ key: String, _ value: Float) {
        VPinballSaveValueFloat(section.rawValue.cstring, key.cstring, value)
    }

    func saveValue(_ section: VPinballSettingsSection, _ key: String, _ value: Bool) {
        VPinballSaveValueBool(section.rawValue.cstring, key.cstring, value ? 1 : 0)
    }

    func saveValue(_ section: VPinballSettingsSection, _ key: String, _ value: String) {
        VPinballSaveValueString(section.rawValue.cstring, key.cstring, value.cstring)
    }

    func rumble(_ data: RumbleData) {
        if data.lowFrequencyRumble > 0 || data.highFrequencyRumble > 0 {
            let style: UIImpactFeedbackGenerator.FeedbackStyle

            if data.lowFrequencyRumble == data.highFrequencyRumble {
                style = .rigid
            } else if data.lowFrequencyRumble > 20000 || data.highFrequencyRumble > 20000 {
                style = .heavy
            } else if data.lowFrequencyRumble > 10000 || data.highFrequencyRumble > 10000 {
                style = .medium
            } else if data.lowFrequencyRumble > 1000 || data.highFrequencyRumble > 1000 {
                style = .light
            } else {
                style = .soft
            }

            impactGenerators[style]?.impactOccurred()
        }
    }

    func play(table: Table) async -> Bool {
        if await MainActor.run(body: { VPinballModel.shared.activeTable != nil }) {
            return false
        }

        if loadValue(.standalone, "ResetLogOnPlay", true) {
            VPinballResetLog()
        }

        await MainActor.run {
            VPinballModel.shared.activeTable = table
            MainViewModel.shared.errorMessage = ""
            tableImageCache.removeAllObjects()

            VPinballModel.shared.showHUD(title: table.name,
                                         status: "Launching")
        }

        var success = true

        if let tablePath = await TableManager.shared.getLoadedTablePath(table: table) {
            if await Task.detached(
                priority: .userInitiated,
                operation: { [tablePath] in
                    VPinballStatus(rawValue: VPinballLoadTable(tablePath.cstring))
                }
            ).value == .success {
                if await MainActor.run(body: { VPinballStatus(rawValue: VPinballPlay()) }) != .success {
                    try? await Task.sleep(nanoseconds: 500_000_000)

                    await MainActor.run { VPinballModel.shared.activeTable = nil }

                    success = false

                    VPinballManager.log(.error, "unable to start table")
                }
            } else {
                try? await Task.sleep(nanoseconds: 500_000_000)

                await MainActor.run { VPinballModel.shared.activeTable = nil }

                success = false

                VPinballManager.log(.error, "unable to load table")
            }

            await MainActor.run {
                VPinballModel.shared.hideHUD()
            }

            return success
        } else {
            try? await Task.sleep(nanoseconds: 500_000_000)
            await MainActor.run {
                VPinballModel.shared.activeTable = nil
                VPinballModel.shared.hideHUD()
            }
            VPinballManager.log(.error, "unable to stage table")
            return false
        }
    }

    func stop() {
        Task { @MainActor in
            if let table = VPinballModel.shared.activeTable {
                TableManager.shared.clearLoadedTable(table: table)
            }
        }
        VPinballStop()
    }

    func resetIni() {
        _ = VPinballResetIni()
    }

    func updateWebServer() {
        VPinballUpdateWebServer()
    }

    func getPath(_ pathType: VPinballPath) -> String {
        return String(cString: VPinballGetPath(pathType.rawValue))
    }
}

File: standalone/android/app/src/main/java/org/vpinball/app/VPinballManager.kt
package org.vpinball.app

import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import android.util.Size
import java.io.File
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import org.koin.core.component.KoinComponent
import org.libsdl.app.SDL
import org.vpinball.app.jni.VPinballCommandData
import org.vpinball.app.jni.VPinballEvent
import org.vpinball.app.jni.VPinballJNI
import org.vpinball.app.jni.VPinballLogLevel
import org.vpinball.app.jni.VPinballPath
import org.vpinball.app.jni.VPinballProgressData
import org.vpinball.app.jni.VPinballRumbleData
import org.vpinball.app.jni.VPinballSettingsSection
import org.vpinball.app.jni.VPinballSettingsSection.STANDALONE
import org.vpinball.app.jni.VPinballStatus
import org.vpinball.app.jni.VPinballWebServerData
import org.vpinball.app.ui.screens.landing.LandingScreenViewModel
import org.vpinball.app.util.FileUtils

object VPinballManager : KoinComponent {
    val vpinballJNI: VPinballJNI = VPinballJNI()

    private lateinit var context: Context
    private lateinit var cacheDir: File
    private lateinit var displaySize: Size
    private lateinit var vibrator: Vibrator

    private var playerActivity: VPinballPlayerActivity? = null
    private var mainActivity: VPinballActivity? = null

    private var lastProgressEvent: VPinballEvent? = null
    private var lastProgress: Int? = null

    enum class InitState {
        NOT_INITIALIZED,
        SDL_READY,
        INITIALIZED,
    }

    private var initState = InitState.NOT_INITIALIZED
    private val initLock = Object()
    private val pendingCallbacks = mutableListOf<() -> Unit>()

    fun initialize(context: Context) {
        this.context = context.applicationContext
        cacheDir = context.cacheDir

        val displayMetrics = context.resources.displayMetrics
        val width = displayMetrics.widthPixels
        val height = displayMetrics.heightPixels
        displaySize = if (width > height) Size(height, width) else Size(width, height)

        vibrator =
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                val vibratorManager = context.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
                vibratorManager.defaultVibrator
            } else {
                context.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
            }

        SAFFileSystem.initialize(context)
    }

    fun onActivityReady(activity: VPinballActivity) {
        synchronized(initLock) {
            mainActivity = activity
            SDL.setContext(activity)

            if (initState != InitState.NOT_INITIALIZED) {
                return
            }

            initState = InitState.SDL_READY
            performInit()
        }
    }

    private fun performInit() {
        vpinballJNI.VPinballInit { value, jsonData ->
            val activity = mainActivity ?: return@VPinballInit
            val viewModel = activity.viewModel
            val event = VPinballEvent.entries.find { it.ordinal == value } ?: return@VPinballInit

            when (event) {
                VPinballEvent.INIT_COMPLETE -> {
                    CoroutineScope(Dispatchers.IO).launch {
                        runCatching { FileUtils.copyAssets(context.assets, "", File(vpinballJNI.VPinballGetPath(VPinballPath.ROOT.value))) }

                        synchronized(initLock) {
                            initState = InitState.INITIALIZED
                            pendingCallbacks.forEach { callback -> CoroutineScope(Dispatchers.Main).launch { callback() } }
                            pendingCallbacks.clear()
                        }
                    }
                }
                VPinballEvent.EXTRACT_SCRIPT,
                VPinballEvent.LOADING_ITEMS,
                VPinballEvent.LOADING_SOUNDS,
                VPinballEvent.LOADING_IMAGES,
                VPinballEvent.LOADING_FONTS,
                VPinballEvent.LOADING_COLLECTIONS,
                VPinballEvent.PRERENDERING -> {
                    val progressData =
                        jsonData?.let { jsonStr ->
                            try {
                                Json.decodeFromString<VPinballProgressData>(jsonStr)
                            } catch (e: Exception) {
                                log(VPinballLogLevel.WARN, "Failed to parse progress data JSON: $jsonStr - ${e.message}")
                                null
                            }
                        }

                    val shouldUpdate = lastProgressEvent != event || lastProgress != progressData?.progress
                    if (shouldUpdate) {
                        log(VPinballLogLevel.INFO, "event=${event.name}, data=${progressData}")
                        lastProgressEvent = event
                        lastProgress = progressData?.progress

                        progressData?.let { CoroutineScope(Dispatchers.Main).launch { viewModel.updateHUD(progressData.progress, event.text ?: "") } }
                    }
                }
                VPinballEvent.PLAYER_STARTED -> {
                    lastProgressEvent = null
                    lastProgress = null
                    CoroutineScope(Dispatchers.Main).launch {
                        viewModel.isPlaying = true
                        delay(500)
                        viewModel.hideHUD()
                    }
                }
                VPinballEvent.RUMBLE -> {
                    val rumbleData =
                        jsonData?.let { jsonStr ->
                            try {
                                Json.decodeFromString<VPinballRumbleData>(jsonStr)
                            } catch (e: Exception) {
                                log(VPinballLogLevel.WARN, "Failed to parse rumble data JSON: $jsonStr - ${e.message}")
                                null
                            }
                        }
                    rumbleData?.let { rumble(it) }
                }
                VPinballEvent.PLAYER_CLOSED -> {
                    val tableToCleanup = viewModel.activeTable
                    viewModel.activeTable = null
                    CoroutineScope(Dispatchers.Main).launch {
                        viewModel.isPlaying = false
                        viewModel.hideHUD()
                        delay(100)

                        tableToCleanup?.let { table ->
                            if (SAFFileSystem.isUsingSAF()) {
                                viewModel.showHUD(table.name, "Saving changes...")
                                delay(50)

                                withContext(Dispatchers.IO) {
                                    TableManager.getInstance().cleanupLoadedTable(table) { progress, status ->
                                        CoroutineScope(Dispatchers.Main).launch { viewModel.updateHUD(progress, status) }
                                    }
                                }

                                viewModel.hideHUD()
                            } else {
                                withContext(Dispatchers.IO) { TableManager.getInstance().cleanupLoadedTable(table) }
                            }

                            withContext(Dispatchers.IO) {
                                TableManager.getInstance().reloadTableImage(table)?.let { updatedTable ->
                                    LandingScreenViewModel.triggerUpdateTable(updatedTable)
                                }
                            }
                        }

                        delay(2000)
                        playerActivity?.finish()
                    }
                }
                VPinballEvent.WEB_SERVER -> {
                    val webServerData =
                        jsonData?.let { jsonStr ->
                            try {
                                Json.decodeFromString<VPinballWebServerData>(jsonStr)
                            } catch (_: Exception) {
                                null
                            }
                        }
                    CoroutineScope(Dispatchers.Main).launch { viewModel.webServerURL = webServerData?.url }
                }
                VPinballEvent.COMMAND -> {
                    val commandData =
                        jsonData?.let { jsonStr ->
                            try {
                                Json.decodeFromString<VPinballCommandData>(jsonStr)
                            } catch (e: Exception) {
                                log(VPinballLogLevel.WARN, "Failed to parse command data JSON: $jsonStr - ${e.message}")
                                null
                            }
                        }
                    commandData?.let {
                        if (it.command == "reloadTables") {
                            CoroutineScope(Dispatchers.IO).launch {
                                TableManager.getInstance().refresh()
                                LandingScreenViewModel.triggerRefresh()
                            }
                        }
                    }
                }
                else -> {
                    log(VPinballLogLevel.WARN, "event=${event}")
                }
            }
        }
    }

    fun whenReady(callback: () -> Unit) {
        synchronized(initLock) {
            if (initState == InitState.INITIALIZED) {
                callback()
            } else {
                pendingCallbacks.add(callback)
            }
        }
    }

    fun isInitialized(): Boolean = synchronized(initLock) { initState == InitState.INITIALIZED }

    fun setPlayerActivity(activity: VPinballPlayerActivity?) {
        playerActivity = activity
    }

    fun setMainActivity(activity: VPinballActivity?) {
        synchronized(initLock) {
            mainActivity = activity
            if (activity != null) {
                SDL.setContext(activity)
            }
        }
    }

    fun getDisplaySize(): Size {
        return displaySize
    }

    fun getCacheDir(): File {
        return cacheDir
    }

    fun log(level: VPinballLogLevel, message: String) {
        vpinballJNI.VPinballLog(level.value, message)
    }

    private fun rumble(data: VPinballRumbleData) {
        if (data.lowFrequencyRumble > 0 || data.highFrequencyRumble > 0) {
            val amplitude =
                when {
                    data.lowFrequencyRumble == data.highFrequencyRumble -> VibrationEffect.DEFAULT_AMPLITUDE
                    data.lowFrequencyRumble > 20000 || data.highFrequencyRumble > 20000 -> 255
                    data.lowFrequencyRumble > 10000 || data.highFrequencyRumble > 10000 -> 200
                    data.lowFrequencyRumble > 1000 || data.highFrequencyRumble > 1000 -> 100
                    else -> 50
                }
            vibrator.vibrate(VibrationEffect.createOneShot(15, amplitude))
        }
    }

    fun updateWebServer() {
        vpinballJNI.VPinballUpdateWebServer()
    }

    fun getVersionString(): String = vpinballJNI.VPinballGetVersionStringFull()

    fun loadValue(section: VPinballSettingsSection, key: String, defaultValue: Int): Int =
        vpinballJNI.VPinballLoadValueInt(section.value, key, defaultValue)

    fun loadValue(section: VPinballSettingsSection, key: String, defaultValue: Float): Float =
        vpinballJNI.VPinballLoadValueFloat(section.value, key, defaultValue)

    fun loadValue(section: VPinballSettingsSection, key: String, defaultValue: Boolean): Boolean =
        vpinballJNI.VPinballLoadValueBool(section.value, key, defaultValue)

    fun loadValue(section: VPinballSettingsSection, key: String, defaultValue: String): String =
        vpinballJNI.VPinballLoadValueString(section.value, key, defaultValue)

    fun saveValue(section: VPinballSettingsSection, key: String, value: Int) {
        vpinballJNI.VPinballSaveValueInt(section.value, key, value)
    }

    fun saveValue(section: VPinballSettingsSection, key: String, value: Float) {
        vpinballJNI.VPinballSaveValueFloat(section.value, key, value)
    }

    fun saveValue(section: VPinballSettingsSection, key: String, value: Boolean) {
        vpinballJNI.VPinballSaveValueBool(section.value, key, value)
    }

    fun saveValue(section: VPinballSettingsSection, key: String, value: String) {
        vpinballJNI.VPinballSaveValueString(section.value, key, value)
    }

    fun resetIni() {
        vpinballJNI.VPinballResetIni()
    }

    suspend fun load(table: Table, onProgress: ((Int, String) -> Unit)? = null): Boolean {
        val viewModel = mainActivity?.viewModel ?: return false

        return withContext(Dispatchers.IO) {
            if (loadValue(STANDALONE, "ResetLogOnPlay", true)) {
                vpinballJNI.VPinballResetLog()
            }

            val tablePath = TableManager.getInstance().stageTable(table) { progress, status -> onProgress?.invoke(progress, status) }

            if (tablePath == null) {
                log(VPinballLogLevel.ERROR, "Unable to stage table: ${table.uuid}")
                delay(500)
                withContext(Dispatchers.Main) {
                    viewModel.activeTable = null
                    viewModel.hideHUD()
                    LandingScreenViewModel.triggerError("Unable to stage table.")
                }
                return@withContext false
            }

            if (vpinballJNI.VPinballLoadTable(tablePath) == VPinballStatus.SUCCESS.value) {
                true
            } else {
                log(VPinballLogLevel.ERROR, "Unable to load table: ${table.uuid}")
                delay(500)
                withContext(Dispatchers.Main) {
                    viewModel.activeTable = null
                    viewModel.hideHUD()
                    LandingScreenViewModel.triggerError("Unable to load table.")
                }
                false
            }
        }
    }

    fun play() {
        if (mainActivity?.viewModel?.activeTable == null) {
            log(VPinballLogLevel.ERROR, "No table loaded for playback")
            return
        }

        vpinballJNI.VPinballPlay()
    }

    fun stop() {
        vpinballJNI.VPinballStop()
    }

    fun showError(message: String) {
        CoroutineScope(Dispatchers.Main).launch {
            delay(250)
            LandingScreenViewModel.triggerError(message)
        }
    }

    fun getPath(pathType: VPinballPath): String = vpinballJNI.VPinballGetPath(pathType.value)
}
